### ISSUE TITLE
improvements of `fold` and `bindings` in tree

### ISSUE THREAD

**Issue by Ngoguey42** [mirage/irmin#1522]
Both `Tree.fold` and `Tree.bindings` call `to_map` when the tag of the node is `Value`, this could be avoided.

`Tree.fold` with ``force = `False`` skips the nodes with tag `Value` when `cached_map t` is `None`. It might be a bug.

**Comment by samoht** [comment]
Just need to be careful here as we might like to keep the semantics of `fold` independent of the caching strategy: e.g. it would be a bit confusing if `Tree.fold f x; Tree.save x; Tree.fold f x`  would give different results. If that's the case we need to specify very clearly that the order in which we visit the children is unstable.

**Comment by Ngoguey42** [comment]
Yes, some of these changes would modify the children ordering and I'm afraid that #1508 already brings these kind of changes. 

This is not the first time that children ordering gets in the way, I think we should proceed with specifying that the ordering is unstable in functions like `Tree.fold`, `Tree.list`, the new`Tree.seq` and also for the export order.

**Comment by Ngoguey42** [comment]
IINM, what's left to do here is to treat the ``~force:`False ~order:`Undefined`` case of `Tree.fold` when a `cached_value` (or a `cached_portable` is available, using something like this: `Backend.Node.Val.seq ~force:false`.
